File: pkg/policy/policy_test.go
package policy

import "testing"

const minorBumpCommit = `Bumps [github.com/foo/bar](https://github.com/foo/bar) from 1.2.3 to 1.3.0.

---
updated-dependencies:
- dependency-name: github.com/foo/bar
  dependency-version: 1.3.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...
`

const majorBumpCommit = `Bumps [react](https://github.com/facebook/react) from 19.1.0 to 20.0.0.

---
updated-dependencies:
- dependency-name: react
  update-type: version-update:semver-major
...
`

const groupedMinorPatchCommit = `Bumps the aws group with 2 updates.

---
updated-dependencies:
- dependency-name: github.com/aws/aws-sdk-go-v2
  update-type: version-update:semver-minor
- dependency-name: github.com/aws/smithy-go
  update-type: version-update:semver-patch
...
`

const groupedWithMajorCommit = `Bumps the tooling group with 2 updates.

---
updated-dependencies:
- dependency-name: left-pad
  update-type: version-update:semver-patch
- dependency-name: webpack
  update-type: version-update:semver-major
...
`

func TestEvaluate(t *testing.T) {
	tests := map[string]struct {
		config string // depsmate.yml content; empty = defaults
		pr     PR
		want   bool
	}{
		"non-dependabot author denied": {
			pr: PR{
				AuthorLogin: "jianzeng",
				HeadRef:     "dependabot/go_modules/github.com/foo/bar-1.3.0",
				Title:       "chore: bump github.com/foo/bar from 1.2.3 to 1.3.0",
			},
			want: false,
		},
		"non-dependabot branch denied": {
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "feature/manual-bump",
				Title:       "bump foo from 1.2.3 to 1.3.0",
			},
			want: false,
		},
		"github actions major bump allowed": {
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/github_actions/actions/checkout-5",
				Title:       "ci: bump actions/checkout from 4 to 5",
			},
			want: true,
		},
		"library minor bump allowed": {
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/github.com/foo/bar-1.3.0",
				Title:             "chore: bump github.com/foo/bar from 1.2.3 to 1.3.0",
				HeadCommitMessage: minorBumpCommit,
			},
			want: true,
		},
		"library major bump denied": {
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/npm_and_yarn/react-20.0.0",
				Title:             "chore: bump react from 19.1.0 to 20.0.0",
				HeadCommitMessage: majorBumpCommit,
			},
			want: false,
		},
		"grouped minor and patch allowed": {
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/aws-1b2c3d",
				Title:             "chore: bump the aws group with 2 updates",
				HeadCommitMessage: groupedMinorPatchCommit,
			},
			want: true,
		},
		"grouped containing major denied": {
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/npm_and_yarn/tooling-4e5f6a",
				Title:             "chore: bump the tooling group with 2 updates",
				HeadCommitMessage: groupedWithMajorCommit,
			},
			want: false,
		},
		"no metadata, title within same major allowed": {
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/pip/requests-2.32.5",
				Title:       "build: bump requests from 2.32.4 to 2.32.5",
			},
			want: true,
		},
		"no metadata, title crossing major denied": {
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/pip/numpy-3.0.0",
				Title:       "build: bump numpy from v2.3.1 to v3.0.0",
			},
			want: false,
		},
		"no metadata, grouped title denied": {
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/npm_and_yarn/tooling-4e5f6a",
				Title:       "chore: bump the tooling group with 3 updates",
			},
			want: false,
		},
		"config disabled denies everything": {
			config: "enabled: false\n",
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/github.com/foo/bar-1.3.0",
				Title:             "chore: bump github.com/foo/bar from 1.2.3 to 1.3.0",
				HeadCommitMessage: minorBumpCommit,
			},
			want: false,
		},
		"config wildcard patch denies minor": {
			config: "ecosystems:\n  \"*\": patch\n",
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/github.com/foo/bar-1.3.0",
				Title:             "chore: bump github.com/foo/bar from 1.2.3 to 1.3.0",
				HeadCommitMessage: minorBumpCommit,
			},
			want: false,
		},
		"config ecosystem major allows major": {
			config: "ecosystems:\n  npm_and_yarn: major\n",
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/npm_and_yarn/react-20.0.0",
				Title:             "chore: bump react from 19.1.0 to 20.0.0",
				HeadCommitMessage: majorBumpCommit,
			},
			want: true,
		},
		"config wildcard merges over default, actions keep major": {
			config: "ecosystems:\n  \"*\": patch\n",
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/github_actions/actions/checkout-5",
				Title:       "ci: bump actions/checkout from 4 to 5",
			},
			want: true,
		},
		"config ecosystem override caps actions at minor": {
			config: "ecosystems:\n  github_actions: minor\n",
			pr: PR{
				AuthorLogin: "dependabot[bot]",
				HeadRef:     "dependabot/github_actions/actions/checkout-5",
				Title:       "ci: bump actions/checkout from 4 to 5",
			},
			want: false,
		},
		"config ignore glob denies matching dependency": {
			config: "ignore:\n  - github.com/aws/*\n",
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/aws-1b2c3d",
				Title:             "chore: bump the aws group with 2 updates",
				HeadCommitMessage: groupedMinorPatchCommit,
			},
			want: false,
		},
		"config ignore exact name denies": {
			config: "ignore:\n  - github.com/foo/bar\n",
			pr: PR{
				AuthorLogin:       "dependabot[bot]",
				HeadRef:           "dependabot/go_modules/github.com/foo/bar-1.3.0",
				Title:             "chore: bump github.com/foo/bar from 1.2.3 to 1.3.0",
				HeadCommitMessage: minorBumpCommit,
			},
			want: false,
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			cfg := Default()
			if tc.config != "" {
				var err error
				cfg, err = ParseConfig([]byte(tc.config))
				if err != nil {
					t.Fatalf("ParseConfig() error: %v", err)
				}
			}
			got := Evaluate(cfg, tc.pr)
			if got.Merge != tc.want {
				t.Errorf("Evaluate() merge = %v, want %v (reason: %s)", got.Merge, tc.want, got.Reason)
			}
			if got.Reason == "" {
				t.Error("Evaluate() returned an empty reason")
			}
		})
	}
}

func TestParseConfigRejectsInvalid(t *testing.T) {
	tests := map[string]string{
		"unknown key":        "max_update_type: minor\n",
		"bad ecosystem type": "ecosystems:\n  pip: everything\n",
		"not yaml":           ":\n  - {",
	}
	for name, content := range tests {
		t.Run(name, func(t *testing.T) {
			if _, err := ParseConfig([]byte(content)); err == nil {
				t.Errorf("ParseConfig(%q) succeeded, want error", content)
			}
		})
	}
}

File: test/policy.test.ts
import { describe, expect, it } from "vitest";
import { defaultConfig, evaluate, parseConfig } from "../src/policy";
import cases from "./policy-cases.json";

describe("policy parity with Go", () => {
  for (const c of cases) it(c.name, () => {
    const result = evaluate(parseConfig(c.config), c.pr);
    expect(result.merge).toBe(c.want);
    expect(result.reason).not.toBe("");
  });
});
it.each(["max_update_type: minor", "ecosystems:\n  pip: everything", ":\n  - {", "enabled: yes", "ignore: foo", "ecosystems: []", "enabled: true\nenabled: false", "ecosystems:\n  '*': constructor"])("rejects invalid config %s", text => {
  expect(() => parseConfig(text)).toThrow();
});
it("rejects untyped non-indirect entries in a group", () => {
  expect(evaluate(defaultConfig(), {
    authorLogin: "dependabot[bot]", headRef: "dependabot/npm_and_yarn/group", title: "Bump a from 1.0 to 1.1",
    headCommitMessage: "- dependency-name: a\n  update-type: version-update:semver-minor\n- dependency-name: b",
  }).merge).toBe(false);
});
it("denies an ecosystem without a fallback", () => {
  const config = defaultConfig(); config.ecosystems = {};
  expect(evaluate(config, { authorLogin: "dependabot[bot]", headRef: "dependabot/npm_and_yarn/foo", title: "Bump foo from 1.0 to 1.1" }).merge).toBe(false);
});

const securityPR = {
  authorLogin: "dependabot[bot]",
  headRef: "dependabot/go_modules/golang.org/x/net-0.55.0",
  title: "build(deps): bump golang.org/x/net from 0.49.0 to 0.55.0",
  headCommitMessage: "---\nupdated-dependencies:\n- dependency-name: golang.org/x/net\n  dependency-version: 0.55.0\n  dependency-type: indirect\n...",
};
it("accepts indirect metadata without update-type", () => {
  expect(evaluate(defaultConfig(), securityPR).merge).toBe(true);
});
it("allows indirect updates regardless of title or version cap", () => {
  expect(evaluate(parseConfig("ecosystems:\n  '*': patch"), {
    ...securityPR, title: "security update with no version information",
  }).merge).toBe(true);
});
it.each(["ignore:\n  - golang.org/x/net", "enabled: false"])("still applies %s", config => {
  expect(evaluate(parseConfig(config), securityPR).merge).toBe(false);
});
it.each(["unknown", "", "version-update:semver-major"])("allows indirect updates with update-type %s", type => {
  expect(evaluate(defaultConfig(), { ...securityPR,
    headCommitMessage: securityPR.headCommitMessage + `\n  update-type: ${type}`,
  }).merge).toBe(true);
});
it.each(["before", "after"])("checks typed entries %s an untyped entry", order => {
  const typed = "- dependency-name: other\n  update-type: version-update:semver-major\n";
  const untyped = "- dependency-name: golang.org/x/net\n  dependency-type: indirect\n";
  expect(evaluate(defaultConfig(), { ...securityPR,
    headCommitMessage: order === "before" ? typed + untyped : untyped + typed,
  }).merge).toBe(false);
});
it("applies ignore rules to every untyped entry in a group", () => {
  expect(evaluate(parseConfig("ignore: [other]"), { ...securityPR,
    headCommitMessage: "- dependency-name: golang.org/x/net\n  dependency-type: indirect\n- dependency-name: other\n  dependency-type: indirect\n",
  }).merge).toBe(false);
});

it.each(["direct:production", "direct:development", "unknown", ""])("requires update-type for %s dependencies", dependencyType => {
  expect(evaluate(defaultConfig(), { ...securityPR,
    headCommitMessage: securityPR.headCommitMessage.replace("dependency-type: indirect", `dependency-type: ${dependencyType}`),
  }).merge).toBe(false);
});
it("allows a mixed group of indirect and direct minor updates", () => {
  expect(evaluate(defaultConfig(), { ...securityPR,
    headCommitMessage: "- dependency-name: indirect/package\n  dependency-type: indirect\n- dependency-name: direct/package\n  dependency-type: direct:production\n  update-type: version-update:semver-minor\n",
  }).merge).toBe(true);
});
